health.go

```diff
@@ -0,0 +1,169 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tchannel
+
+import (
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+)
+
+const (
+	_defaultHealthCheckTimeout         = time.Second
+	_defaultHealthCheckFailuresToClose = 5
+
+	_healthHistorySize = 256
+)
+
+// HealthCheckOptions are the parameters to configure active TChannel health
+// checks. These are not intended to check application level health, but
+// TCP connection health (similar to TCP keep-alives). The health checks use
+// TChannel ping messages.
+type HealthCheckOptions struct {
+	// The period between health checks. If this is zeor, active health checks
+	// are disabled.
+	Interval time.Duration
+
+	// The timeout to use for a health check.
+	// If no value is specified, it defaults to time.Second.
+	Timeout time.Duration
+
+	// FailuresToClose is the number of consecutive health check failures that
+	// will cause this connection to be closed.
+	// If no value is specified, it defaults to 5.
+	FailuresToClose int
+}
+
+type healthHistory struct {
+	sync.RWMutex
+
+	states []bool
+
+	insertAt int
+	total    int
+}
+
+func newHealthHistory() *healthHistory {
+	return &healthHistory{
+		states: make([]bool, _healthHistorySize),
+	}
+}
+
+func (hh *healthHistory) add(b bool) {
+	hh.Lock()
+	defer hh.Unlock()
+
+	hh.states[hh.insertAt] = b
+	hh.insertAt = (hh.insertAt + 1) % _healthHistorySize
+	hh.total++
+}
+
+func (hh *healthHistory) asBools() []bool {
+	hh.RLock()
+	defer hh.RUnlock()
+
+	if hh.total < _healthHistorySize {
+		return append([]bool(nil), hh.states[:hh.total]...)
+	}
+
+	states := hh.states
+	copyStates := make([]bool, 0, _healthHistorySize)
+	copyStates = append(copyStates, states[hh.insertAt:]...)
+	copyStates = append(copyStates, states[:hh.insertAt]...)
+	return copyStates
+}
+
+func (hco HealthCheckOptions) enabled() bool {
+	return hco.Interval > 0
+}
+
+func (hco HealthCheckOptions) withDefaults() HealthCheckOptions {
+	if hco.Timeout == 0 {
+		hco.Timeout = _defaultHealthCheckTimeout
+	}
+	if hco.FailuresToClose == 0 {
+		hco.FailuresToClose = _defaultHealthCheckFailuresToClose
+	}
+	return hco
+}
+
+// healthCheck will do periodic pings on the connection to check the state of the connection.
+// We accept connID on the stack so can more easily debug panics or leaked goroutines.
+func (c *Connection) healthCheck(connID uint32) {
+	opts := c.opts.HealthChecks
+
+	ticker := c.timeTicker(opts.Interval)
+	defer ticker.Stop()
+
+	consecutiveFailures := 0
+	for {
+		select {
+		case <-ticker.C:
+		case <-c.healthCheckCtx.Done():
+			return
+		}
+
+		ctx, cancel := context.WithTimeout(c.healthCheckCtx, opts.Timeout)
+		defer cancel()
```

Comment:
Do we want to defer this until the function returns, or should we make sure to execute it at the end of this loop iteration?

Reply:
Good catch! Could have been a huge memory leak, changed ti to `cancel` immediately after the `ping` method, since we don't use the `ctx` for anything else.